src/conn/pipeline.rs

```diff
@@ -0,0 +1,80 @@
+use mysql_common::proto::codec::PacketCodecState;
+use std::{collections::VecDeque, marker::PhantomData};
+
+use crate::{
+    conn::{query_result::Binary, ConnMut},
+    prelude::*,
+    Conn, Params, QueryResult, Result, Statement,
+};
+
+struct PipelineInner<'c> {
+    conn: &'c mut Conn,
+    cmd_states: VecDeque<PacketCodecState>,
+}
+
+impl<'c> PipelineInner<'c> {
+    fn exec(&mut self, stmt: &Statement, params: Params) -> Result<()> {
+        self.conn._execute_pipeline(stmt, params)?;
+        self.cmd_states
+            .push_back(self.conn.stream_mut().codec().save_state());
+        Ok(())
+    }
+
+    fn finish(self) -> PipelineResult<'c, Binary> {
+        PipelineResult {
+            conn: self.conn,
+            cmd_states: self.cmd_states,
+            protocol: PhantomData,
+        }
+    }
+}
+
+pub struct Pipeline<'c>(Option<PipelineInner<'c>>);
```

Comment:
Why are we wrapping `PipelineInner` in an `Option`?

Reply:
This is pretty subtle, but it's necessary because `Pipeline` implements  `Drop`.

Normally when you have an owned value, you can move sub-fields out of it, as long as you don't try to pass that original value somewhere else (by reference or otherwise) after moving a field out. This means you can write a function like `struct Foo<T>(T); fn bar<T>(foo: Foo<T>) -> T { foo.0 }`, which is a fairly common pattern (e.g. `.into` and `.unwrap` functions).

But in that example, if `Foo` implements `Drop` (which takes `&mut self`), there's a problem. `bar` owns the parameter `foo`, so it needs to get dropped at the end of the function. If Rust calls `drop` after moving out `.0`, it's passing a reference to a value that has had a sub-field moved out of it. If it calls `drop` before moving out `.0`, it's accessing a value that has already been dropped. Neither is allowed, so Rust prevents moving out of a type that implements `Drop`. 

Another way that you can take ownership of an object is by swapping it with another object that you own. For example, this function is perfectly legal even if `Foo` implements `Drop`: `fn baz<T>(mut foo: Foo<T>, mut t: T) { std::mem::swap(&mut foo, &mut T); t }`. This is fine because nothing ever gets moved out of `foo`; the contents just get exchanged with a new value. When `drop` runs at the end of the function, it sees whatever value was swapped in.

The `finish` function on `Pipeline` takes ownership of `self`, and needs to move the connection reference out so it can pass it to `PipelineResult`. But `Pipeline` implements `Drop`, so we can't do that. We also can't swap in another value because we have no way to create a `&'c mut Conn` from thin air. The solution is to wrap it in an `Option`, so we can swap in `None`, which is exactly what `.take` does! So as `finish` returns, `drop` gets called with a `None` value, and it skips the cleanup because the reference has already been handed off to the result object. If the `Pipeline` gets dropped _before_ `finish` is called, then the `Drop` implementation will have a live reference and can do the cleanup.